monai/losses/focal_loss.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2020 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.autograd import Variable
+
+
+class FocalLoss(nn.Module):
+    """
+    PyTorch implementation of the Focal Loss.
+    [1] "Focal Loss for Dense Object Detection", T. Lin et al., ICCV 2017
+    """
+    def __init__(self, gamma=2., alpha=None, reduction='mean'):
+        """
+        Args:
+            gamma: (float) value of the exponent gamma in the definition
+            of the Focal loss.
+            alpha: (float or float list or None) weights to apply to the
```

Comment:
I would suggest describing alpha in a little more detail, what shape should it have it not a single float? If the loss is being used as a categorical loss for a classifier, there's the NxC prediction tensor for N instances with C classes, should alpha then be shape N? If I want to weight each class then do I need to compute alpha for each minibatch, assigning the weight corresponding to the ground truth class assignment at each instance N?

Reply:
Thank you @ericspod for taking the time to review the code!

In the Focal loss paper alpha is static (it does not depend on the iteration or the mini-batch) and is of shape C. 

In fact, alpha is exactly the same as the weights of a weighted cross entropy. So it is probably best to align the API with the one of the cross entropy loss of PyTorch https://pytorch.org/docs/stable/_modules/torch/nn/modules/loss.html#CrossEntropyLoss
i.e. the Focal loss would inherit from _WeightedLoss and alpha would be replaced by the parameter weight (which will be a tensor).

Please let me know what you think about this.

Reply:
I understand a bit better now what the implementation here is doing versus [this one](https://github.com/clcarwin/focal_loss_pytorch/blob/master/focalloss.py), and is doing the same thing. An example of using the loss would help (which we need elsewhere as well):

```python
pred=torch.tensor([[1,0],[0,1],[1,0]],dtype=torch.float32)
grnd=torch.tensor([0,1,0],dtype=torch.int64)
fl=FocalLoss()
fl(pred,grnd)
``` 

The `_WeightedLoss` type doesn't add much except the `weight` attribute so it's probably a good idea. In that case as well the `reduction` argument of the `_Loss` constructor should be used as well to set the attribute `reduction` as expected rather than overwriting it.